lib/masamune/abstract_syntax_tree.rb

```diff
@@ -11,145 +11,86 @@ def initialize(code)
         LexNode.new(raw_lex_nodes.index(lex_node), lex_node, self.__id__)
       end
 
-      @node_list = []
-      @data_node_list = []
-      @comment_list = []
-      register_nodes(@tree)
-
-      # Refer to Masamune::AbstractSyntaxTree::Comment
-      # to see why we register these separately.
-      register_comments
+      @prism = Prism.parse(code)
+      @nodes = []
+      register_nodes
+      @nodes = order_nodes(@nodes)
     end
 
-    def register_nodes(tree_node = self.tree)
-      if tree_node.is_a?(Array)
-        klass = get_node_class(tree_node.first)
-        msmn_node = klass.new(tree_node, self.__id__)
-      else
-        # Create a general node if the node is a single value.
-        msmn_node = Node.new(tree_node, self.__id__)
-      end
-
-      # Register nodes and any data nodes housed within it.
-      # See Masamune::AbstractSyntaxTree::DataNode for more details on what a data node is.
-      @node_list << msmn_node
-      msmn_node.data_nodes.each { |dn| @data_node_list << dn } if msmn_node.data_nodes
+    def register_nodes(tree_node = @prism.value)
+      # Include our in-house Masamune node helper for all Prism nodes present.
+      tree_node.class.include(NodeHelper) unless tree_node.class.include?(NodeHelper)
 
-      # Continue down the tree until base case is reached.
-      if !msmn_node.nil? && msmn_node.contents.is_a?(Array)
-        msmn_node.contents.each { |node| register_nodes(node) }
+      @nodes << tree_node
+      tree_node.compact_child_nodes.each do |child_node|
+        register_nodes(child_node)
       end
     end
 
-    def register_comments
-      comment_lex_nodes = @lex_nodes.select {|node| node.type == :comment}.flatten
-      @comment_list << comment_lex_nodes.map {|node| Comment.new(node, self.__id__)}
-      @comment_list.flatten!
+    def find_nodes(node_classes, token: nil)
+      node_classes = Array(node_classes)
+      result = @nodes.select {|node| node_classes.include?(node.class)}
```

Comment:
Right now you're putting every node into a big array and then doing a select here. I would highly suggest trying out the visitor classes instead. It should make this operation quite a bit smoother.

Reply:
Ok, so I've implemented `SearchVisitor`, a class that inherits `Prism::Visitor` and overrides the visit node methods as the [documentation](https://ruby.github.io/prism/Prism/Visitor.html) suggests to do. I added `search_visitor.rb` in this commit: https://github.com/gazayas/masamune-ast/pull/60/commits/f87564168095b3109414f54eae4cb804b2278382

I found that each node has its own respective method for visiting. So `Prism::DefNode` has `visit_def_node`, `Prism::StringNode` has `visit_string_node`, etc.

The `SearchVisitor` class goes through all the necessary nodes that we use for Masamune, overrides them and registers the nodes in question to the search results accordingly.

I'm seeing that using `super` traverses the tree, so I used that to ensure every node is registered. I'm hoping this is the proper way to use this class.